src/main/java/org/sonarqube/gradle/SonarTask.java

```diff
@@ -194,6 +201,201 @@ public Provider<Map<String, String>> getProperties() {
+        LOGGER.debug("Resolving main class path for top-level project.");
+      } else {
+        LOGGER.debug("Resolving main class path for {}.", projectName);
+      }
+    }
+    if (mainClassPath == null) {
+      LOGGER.debug("No main class path configured. Skipping resolution.");
+      return;
+    }
+
+    List<File> resolvedLibraries = SonarUtils.exists(mainClassPath);
+    String resolvedAsAString = resolvedLibraries.stream()
+            .filter(File::exists)
+            .map(File::getAbsolutePath)
+            .collect(Collectors.joining(","));
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Resolved configured main class path as: {}", resolvedAsAString);
+    }
+
+    String propertyKey = isTopLevelProject ?
+            "sonar.java.libraries" :
+            (":" + projectName + ".sonar.java.libraries");
+    String legacyPropertyKey = isTopLevelProject ?
+            "sonar.libraries" :
+            (":" + projectName + ".sonar.libraries");
+
+    String libraries = properties.getOrDefault(propertyKey, "");
+    if (libraries.isEmpty()) {
+      libraries = resolvedAsAString;
+    } else {
+      libraries += "," + resolvedAsAString;
+    }
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Resolved {} as: {}", propertyKey, libraries);
+    }
+
+    properties.put(propertyKey, libraries);
+    properties.put(legacyPropertyKey, libraries);
+  }
+
+  /**
+   * Complete the resolution of <pre>sonar.java.test.libraries</pre> by combining:
+   * <ol>
+   *   <li><pre>sonar.java.binaries</pre></li>
+   *   <li>the existing property</li>
+   *   <li>the resolution of the file collection provided at configuration time</li>
+   * </ol>
+   * The end result is stored in the map passed as input.
+   */
+  static void resolveSonarJavaTestLibraries(String projectName, @Nullable Iterable<File> testClassPath, Map<String, String> properties) {
+    boolean isTopLevelProject = projectName.isBlank();
+    if (LOGGER.isDebugEnabled()) {
+      if (isTopLevelProject) {
+        LOGGER.debug("Resolving test class path for top-level project.");
+      } else {
+        LOGGER.debug("Resolving test class path for {}.", projectName);
+      }
+    }
+
+    if (testClassPath == null) {
+      LOGGER.debug("No test class path configured. Skipping resolution.");
+      return;
+    }
+
+    List<File> resolvedLibraries = SonarUtils.exists(testClassPath);
+    String resolvedAsAString = resolvedLibraries.stream()
+            .filter(File::exists)
+            .map(File::getAbsolutePath)
+            .collect(Collectors.joining(","));
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Resolved configured test class path as: {}", resolvedAsAString);
+    }
+
+    // Prepend sonar.java.binaries if it exists
+    String binariesPropertyKey = isTopLevelProject ?
+            "sonar.java.binaries" :
+            (":" + projectName + ".sonar.java.binaries");
+    String libraries = properties.getOrDefault(binariesPropertyKey, "");
```

Comment:
I see that we do not do this for the main classpath, is this correct?

Reply:
That is correct. If we look at the properties dumped with previous versions of the analyzer, the binaries are always pre-pended to `sonar.java.test.libraries`.